macros/src/parser/lifetimes.rs

```diff
@@ -0,0 +1,98 @@
+use quote::quote;
+use std::ops::Deref;
+use syn::{parse, spanned::Spanned, GenericArgument, Lifetime, PathArguments, Type};
+
+#[derive(Debug, Clone)]
+pub struct Lifetimes(Vec<Lifetime>);
+
+impl Lifetimes {
+    pub fn new() -> Lifetimes {
+        Lifetimes(Vec::new())
+    }
+
+    pub fn from_type(data_type: &Type) -> Result<Lifetimes, parse::Error> {
+        let mut lifetimes = Lifetimes::new();
+        get_lifetimes(data_type, &mut lifetimes)?;
+        Ok(lifetimes)
+    }
+
+    pub fn push(&mut self, lifetime: &Lifetime) {
+        if !self.0.contains(lifetime) {
+            self.0.push(lifetime.to_owned());
+        }
+    }
+
+    pub fn append(&mut self, lifetimes: &Lifetimes) {
+        for lifetime in lifetimes.0.iter() {
+            self.push(lifetime);
+        }
+    }
+
+    pub fn remove(&mut self, lifetimes: &Lifetimes) {
+        self.0.retain(|lt| !lifetimes.0.contains(lt));
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn to_code(&self) -> proc_macro2::TokenStream {
+        if self.is_empty() {
+            quote! {}
+        } else {
+            let lifetimes = &self.0;
+            quote! { #(#lifetimes),* ,}
+        }
+    }
+}
+
+impl Default for Lifetimes {
+    fn default() -> Self {
+        Lifetimes::new()
+    }
+}
+
+impl Deref for Lifetimes {
+    type Target = [Lifetime];
+
+    fn deref(&self) -> &[Lifetime] {
+        &self.0[..]
+    }
+}
+
+// helper function for extracting lifetimes from a Type
+fn get_lifetimes(data_type: &Type, lifetimes: &mut Lifetimes) -> Result<(), parse::Error> {
+    match data_type {
+        Type::Reference(tr) => {
+            if let Some(lifetime) = &tr.lifetime {
+                lifetimes.push(lifetime);
+            } else {
+                return Err(parse::Error::new(
+                    data_type.span(),
+                    "This event's data lifetime is not defined, consider adding a lifetime.",
+                ));
+            }
+            Ok(())
+        }
+        Type::Path(tp) => {
+            let punct = &tp.path.segments;
+            for p in punct.iter() {
+                if let PathArguments::AngleBracketed(abga) = &p.arguments {
+                    for arg in &abga.args {
+                        if let GenericArgument::Lifetime(lifetime) = &arg {
+                            lifetimes.push(lifetime);
+                        }
+                    }
+                }
+            }
+            Ok(())
+        }
+        Type::Tuple(tuple) => {
+            for elem in tuple.elems.iter() {
+                get_lifetimes(elem, lifetimes)?;
+            }
+            Ok(())
+        }
+        _ => Ok(()),
```

Comment:
For all these branches, just let them return nothing and then at the end add a single `Ok()` on the final line of the function.